### ISSUE TITLE
[UI][Console View] Previous changes converted into unknown revision (a few seconds ago)

### ISSUE THREAD

**Issue by Frodox** [buildbot/buildbot#3363]
I love Console view plugin a lot. It is what I need for my workflow (with some improvements).

But I see noisy bug. All is fine when I just started bb and have <=25 changes (I use GitLab web hook only, no fake changes). But when I have 26, 27, 28 ++ changes, first one (1, 2, 3) becomes 'unknown revision'. But in fact it was fine, and build logs show it is fine. I think previous changes should be just hidden and not pull-ed in.
But in best way I want to increase list if changes from 25 to 50, 100, and have pagination for old one.
![selection_011](https://user-images.githubusercontent.com/3653069/27477594-48e5a15c-5815-11e7-8ac8-7dc0f329b166.png)
![selection_012](https://user-images.githubusercontent.com/3653069/27477595-48e5dde8-5815-11e7-9675-fb8381c5a7ca.png)
![selection_013](https://user-images.githubusercontent.com/3653069/27477596-48ee5090-5815-11e7-8424-4b1bfde310cf.png)

**Comment by estan** [comment]
I think we're suffering from this as well. For example:

![console](https://user-images.githubusercontent.com/86675/29994194-18ef346c-8fca-11e7-8864-4a9725677501.png)

**Comment by tardyp** [comment]
You can increase the number of fetched changes in the settings page. We
should eventually find an heuristique to automatically fetch me changes.

**Comment by estan** [comment]
@tardyp: Ah,  but the problem is not the number of changes (that's OK for me), but that revisions are shown as "unknown".

**Comment by Frodox** [comment]
@estan do you use ForceScheduler also?

**Comment by estan** [comment]
@Frodox Yes, not sure if all those unknown revisions in my screenshot are forced builds or not though.

But in any case, surely Buildbot should be able to get the info about the `got_revision` (the one that was actually fetched by the forced build) and present that?

**Comment by s0undt3ch** [comment]
We use the force scheduler but we have way more unknown revisions the the number of forced builds we triggered.
We do however base most triggering on GH events and not polling...

**Comment by tardyp** [comment]
So it looks like there is an issue with the UI decoding the got_revision value. Anybody want to debug this?

This needs to set a breakpoint here:
https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee#L270

**Comment by estan** [comment]
Sorry for the late reply, I followed the quick start for getting a `master` version of the UI running against our master.

I've never developed CoffeeScript (nor JavaScript), and I couldn't get debugging in Firefox working when I launched with `gulp dev proxy` in `www/base`, so couldn't set a breakpoint. I think some mappings between js and coffee are not set up for the `console_view` plugin, only the `base`.

I did some printing in the method you linked though, and I don't think there's anything wrong with the decoding of the `got_revision`. It's just that there's no `SSID -> Change` mapping in `@changesBySSID` for forced builds (they don't even have a `Change`, do they?). So I'm not sure how this could even work for forced builds.

I may have misunderstood how things work though.

**Comment by deamen** [comment]
Same story here, sometimes it is correct, sometimes it's not.
![buildbot_incorrect](https://user-images.githubusercontent.com/15974434/30575361-7eccbdf0-9d43-11e7-81b4-de2927c085e3.png)
![buildbot_correct](https://user-images.githubusercontent.com/15974434/30575364-818163fc-9d43-11e7-8201-328ee85d74c3.png)

**Comment by deamen** [comment]
@estan 
When I run gulp dev proxy in www/base, I don't get console_view, how can I get it?

**Comment by estan** [comment]
@deamen: Hm, I don't think I did anything special except having it turned on in the config of the master I proxied to. Perhaps @tardyp knows?

Maybe you need to run `gulp dev` also in www/console_view? But I don't think that should be necessary to just get it to show..

**Comment by tardyp** [comment]
you need to configure it in the master.cfg

if you want to hack it, you need a gulp dev in it.

**Comment by deamen** [comment]
@tardyp: I still can't get console view on my develop box.
Let's say I want to use the following command to proxy nine.buildbot.net as the  [developer page]( http://docs.buildbot.net/0.9.11/developer/www-base-app.html) instructed in section "Testing with real data":
`gulp dev proxy --host nine.buildbot.net`

When I run the command in www/base, I get all the data but I don't have console view in the GUI.

Where should I put master.cfg? I tried to put it under www/base, it didn't change anything.

@estan: It looks like the issue is gone after I upgraded buildbot from 0.9.9.post2 to 0.9.11, can you confirm?  I see this very often on 0.9.9.post2 when I switch beteewn waterfall_view and console_view, I havn't seen this issue since the upgrade.

**Comment by estan** [comment]
The master.cfg is the master configuration, so you won't find it in your local tree. It's at the buildbot master. So of you're proxying to nine.buildbot.net you'll have no control over how it's configured. I proxied to our own buildbot master when I was debugging.

I will try 0.9.11 (using 0.9.10 now), but having looked at the code I fail to see how it could ever work for forced builds, so I doubt it'll make any difference. Maybe we're not talking about the same problem. But I'll give it a try.

**Comment by estan** [comment]
For me it's not about what happens when switching between views. For me, forced builds are always listed without proper information in the console view. And looking at the code, I think it extracts this information from the corresponding `Change`, but I don't think forced builds have a `Change` (?), so I don't see how it could work. Unless something big happened in 0.9.11. I'm on the bus so a bit hard to have a look right now.

**Comment by tardyp** [comment]
@estan it either extract the info from change, or from the got_revision property, if no change is found.

**Comment by estan** [comment]
@tardyp: Hm, so are you saying it should work? After it has the got_revision, how does it get the information from there?

To me it just looks like it'll enter [here](https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee#L274), where it will call [makeFakeChange](https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee#L284), and in there it tries to fetch a change based on SSID, which will fail for a forced build, so it'll create a fake change with "unknown" author et.c., which is the problem we're seeing.

I still don't see how this could give proper info for forced builds.

**Comment by pohmelie** [comment]
Vote up, our forced builds have no info about commits, repos, codebases (just couple extra properties). So, forced builds are always at top and always "few seconds ago".
![screenshot-2017-10-4 buildbot console view 1](https://user-images.githubusercontent.com/2385765/31191147-f017fb4c-a945-11e7-9e51-7d4d7f46065e.png)

**Comment by pohmelie** [comment]
It looks like [main.module.coffee#L280](https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee#L280) should be changed to
``` coffee
change = @makeFakeChange("unknown codebase", "unknown revision", build.started_at)
```

**Comment by pohmelie** [comment]
Hm, actually this solved the problem of "always «a few seconds ago»", and all this forced builds will go down. But this did not solve, that all forced builds are grouped.
![ff51cce4-c3b5-4be8-9251-6062a59f1a94](https://user-images.githubusercontent.com/2385765/31192564-42d8373a-a94a-11e7-981f-47eba94932d2.png)

**Comment by pohmelie** [comment]
More to say, this does not make force builds go down. Continue digging…

**Comment by pohmelie** [comment]
Ok, I got the problem. Right now console view match build with change. Forced builds have no change. Console view make fake change (**one for all forced builds** without sourcestamps/revisions/etc.). Looks like builds in console view sorted by `changeid`, which is `"unknown revision"` for forced builds. That is why they are always on top and stacked.
I'm not too familiar with angular, but is it hard to sort console view output by `started_at` or something time-like? Cause forced builds definitely have some time.

**Comment by tardyp** [comment]
It should not be hard to fix, but we need somebody to find some time to do
it (and to test it properly with real data)

https://medium.com/buildbot/buildbot-ui-plugin-for-python-developer-ef9dcfdedac0

**Comment by Frodox** [comment]
@tardyp is it possible to map forced build with some 'change' ? Like, specifying correct/real sourcestamp (rev, repository, branch), so console view would understand that this force build match some changes

**Comment by aureliendavid** [comment]
I had the same problem for nightly builds that aren't triggered by a change (when `onlyIfChanged=False`).

I don't have a testing environment either but it seems that it could be done by matching `got_revision` to a change.

In the same way that the list `@changesBySSID` is filled [here](https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee#L107) it's possible to fill a list of `@changesByRevision`

And then [here](https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee#L270) instead of making a fake change based on the `got_revision` it would be possible get the proper change using the new list.

**Comment by estan** [comment]
@aureliendavid What you say sounds hopeful, but are you sure there even is a `Change` for forced builds? Because if there isn't, then it wouldn't matter that you build up a `@changesByRevision`, because there wouldn't be any `Change` in there matching the `got_revision` of the forced build..

**Comment by pohmelie** [comment]
I'm sure, that for my forced builds there is no any `Change`. More to say, it is difficult to make fake change, which will be in right order in console view when sorting is done by `changeid`.

**Comment by pohmelie** [comment]
By the way, where sorting is actually happens? I see only builders sorting in `main.module.coffee`.

**Comment by aureliendavid** [comment]
@estan In my case the `Change` exists because the repo is watched by a gitpoller so every revision has a change associated with it (If I understand it correctly). The got_revision of a forced/nightly build could then be matched to a change. 

But once again I haven't tested this. 

On the console page, in the devtools / network view of a browser, you can see `/changes` api call and check in the response what changes are known to the view at that point.

**Comment by pohmelie** [comment]
@aureliendavid 
Did you test this? Since I have 3 builds (one from git poller, two forced) on my test environment and I have 1 change as response of `/api/v2/changes?limit=30&order=-changeid`

**Comment by aureliendavid** [comment]
@pohmelie what are you building in the forced builds? (another branch that is not watched by the gitpoller for example?) Do they have a got_revision? 

If no change object exists for the got_revision being built, I don't really see what the console could dispay other than a fake change.

**Comment by pohmelie** [comment]
@aureliendavid 
Our workflow for building packages is like this: we have pretty complex cli builder, which invoke docker and run some build process inside for some platform and some target. There is multiple targets and platforms. Some targets autobuilds by commit, but most are just 3rdparty libraries and we update them very rarely. That is why we need force build, where we just pass platform and targets as properties. Actually, some libraries taken from pypi, they just can't have revision/change/etc. in this case.
I've just inspect `build` objects for `got_revision` field, and all 3 builds (1 by commit, 2 forced) have no `properties.got_revision`.

**Comment by pohmelie** [comment]
@aureliendavid 
> I don't really see what the console could dispay other than a fake change.
Yes, that is the problem, I'm ok with fake change displaying, the problem is grouping and ordering. I see solution as change of "group_by" and "order" rules. Group by change id is this exists and order by `started at` or something time-like.

**Comment by aureliendavid** [comment]
OK that's two different issues then. 

1. builds that don't have an `ssid` but do have a `got_revision` should be matched to an actual change if it exists (this should cover forced and timed builds that are on the watched repo)

2. for actual `fakechange` builds:
   * display the revision if known
   * if not known don't group all "unknown revision" together
   * display the actual start time if known
   * sort the whole thing properly


What I was talking about (with the @chagesByRevision list) should work for 1. 

For 2. I don't know... When calling makeFakeChange [here](https://github.com/buildbot/buildbot/blob/master/www/console_view/src/module/main.module.coffee#L280) you can already pass the actual started_at time if know. You could also pass the revision if known but not matched. And to prevent grouping I guess you could generate a unique name instead of "unknown revision" (like "unknown revision N" with N autoincrement or something). As for sorting.. I have no idea.

**Comment by estan** [comment]
If 1 could be worked out and works OK that would be great and would cover our use case.

**Comment by aureliendavid** [comment]
I was able to test for a specific case of 2. (above), it seems to work. See PR details.

**Comment by macdems** [comment]
Just one finding for this issue:

If you specify the revision for the forced build manually (instead of leaving this empty) , the build is able to find the right change.

It looks like BB is using the property `revision` to match the change. The correct behavior would be to try `revision` first, and fallback to `got_revision`.

**Comment by mojca** [comment]
@rajdeepbharati, @p12tic: not sure if this is strictly related, but maybe you can consider it as a data point for testing other tickets related to console view.

**Comment by rajdeepbharati** [comment]
This problems occurs in case of force/nightly builds (not regular builds that are triggered by a commit)

**Comment by dol-sen** [comment]
we are experiencing this "unknown revision" issue in console-view intermittently in our new gemtoo-ci bot.  I just now figured out a trigger for it.   With the page left open, history goes back a long way, was no issue with the long queue of incomplete jobs. When I did a page refresh, the console view history did not go back far enough to include those incomplete jobs.  So the same few jobs that just completed now  had the "unknown revision" bug and were not displayed in the console view original commit line since it was not in the view.

So it seems to be a history issue.  If the originating commit is not in the view scroll, it doesn't have the info to display correctly.

**Comment by dol-sen** [comment]
changing the console-view settings to 1000 buildlimit and 150 chagelimit fixed the display bug for us so far.  Since it is in development only atm, it had a 4 hour backlog of build requests in the queue (only 1 worker so far).  A total of 7 builders, the 2nd last one waiting in build requests, it triggers the last builder to parse the logs, report success/fail.
